src/libraries/System.Private.CoreLib/src/System/Threading/Interlocked.cs

```diff
@@ -168,9 +241,89 @@ public static UIntPtr Exchange(ref UIntPtr location1, UIntPtr value)
+            return CompareExchange(ref location1, value, comparand); // Must expand intrinsic
+#else
+            // this relies on GC keeping 4B alignment for refs and on subtracting to such alignment being in the same object
+            nuint offset = (nuint)Unsafe.AsPointer(ref location1) % sizeof(uint);
+            ref uint alignedRef = ref Unsafe.As<byte, uint>(ref Unsafe.SubtractByteOffset(ref location1, offset));
+            int bitOffset =
+                (int)((BitConverter.IsLittleEndian ? offset : sizeof(uint) - offset - sizeof(byte)) * 8); // to bit offset
+            Debug.Assert(bitOffset is 0 or 8 or 16 or 24);
+            uint mask = ~((uint)byte.MaxValue << bitOffset);
+            uint shiftedValue = (uint)value << bitOffset;
+            uint shiftedComparand = (uint)comparand << bitOffset;
+
+            uint originalValue, fullComparand, newValue;
+            do
+            {
+                // make sure the ref is still aligned
+                Debug.Assert((nuint)Unsafe.AsPointer(ref alignedRef) % sizeof(uint) == 0);
+                originalValue = Volatile.Read(ref alignedRef);
+                uint otherMemory = originalValue & mask;
+                fullComparand = otherMemory | shiftedComparand;
+                newValue = otherMemory | shiftedValue;
+            } while (originalValue != CompareExchange(ref alignedRef, newValue, fullComparand));
+
+            // verify the GC hasn't broken the ref
+            Debug.Assert((nuint)Unsafe.ByteOffset(ref Unsafe.As<uint, byte>(ref alignedRef), ref location1) == offset);
+            return (byte)(originalValue >> bitOffset);
+#endif
+        }
+
+        /// <summary>Compares two 16-bit signed integers for equality and, if they are equal, replaces the first value.</summary>
+        /// <param name="location1">The destination, whose value is compared with <paramref name="comparand"/> and possibly replaced.</param>
+        /// <param name="value">The value that replaces the destination value if the comparison results in equality.</param>
+        /// <param name="comparand">The value that is compared to the value at <paramref name="location1"/>.</param>
+        /// <returns>The original value in <paramref name="location1"/>.</returns>
+        /// <exception cref="NullReferenceException">The address of <paramref name="location1"/> is a null pointer.</exception>
+        [Intrinsic]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         [CLSCompliant(false)]
-        public static ushort CompareExchange(ref ushort location1, ushort value, ushort comparand) =>
-            (ushort)CompareExchange(ref Unsafe.As<ushort, short>(ref location1), (short)value, (short)comparand);
+        public static unsafe ushort CompareExchange(ref ushort location1, ushort value, ushort comparand)
+        {
+#if TARGET_X86 || TARGET_AMD64 || TARGET_ARM64
+            return CompareExchange(ref location1, value, comparand); // Must expand intrinsic
+#else
```

Comment:
I don't think this will work on Mono.   You removed the `byte&` and `int16&` icalls, but the JIT intrinsics handling only works with 32-bit or 64-bit types

https://github.com/dotnet/runtime/blob/81dfdace71c7833c5effa9f696b28efd25d6ee94/src/mono/mono/mini/intrinsics.c#L1452-L1481

Reply:
Ah right I forgot to change the `#if`.